File: code/src/copy.rs
use std::io::{Read, Write};
use std::sync::Arc;
use std::thread;

use crossbeam_channel::{bounded, Receiver, Sender};

use crate::database::Database;
use crate::hash;
use crate::logging::Logger;
use crate::peer::{PeerError, PeerFs};
use crate::pool::ConnectedPeer;
use crate::timestamp;

const CHANNEL_BOUND: usize = 16;

/// A file copy task to be executed.
pub struct CopyTask {
    pub src_peer: String,
    pub dst_peer: String,
    pub path: String,
    pub src_mod_time: String,
}

/// Execute all queued copies concurrently, respecting connection pool limits.
pub fn execute_copies(
    tasks: Vec<CopyTask>,
    peers: &[Arc<ConnectedPeer>],
    db: &Arc<Database>,
    logger: &Arc<Logger>,
    sync_stamp: &str,
) {
    if tasks.is_empty() {
        return;
    }

    let sync_stamp = sync_stamp.to_string();

    thread::scope(|s| {
        let mut handles = Vec::new();
        for task in tasks {
            let peers = peers;
            let db = db.clone();
            let logger = logger.clone();
            let sync_stamp = sync_stamp.clone();

            let src_peer = find_peer(peers, &task.src_peer);
            let dst_peer = find_peer(peers, &task.dst_peer);

            if src_peer.is_none() || dst_peer.is_none() {
                logger.error(&format!("Copy skipped {}: peer not found", task.path));
                continue;
            }

            let src = src_peer.unwrap().clone();
            let dst = dst_peer.unwrap().clone();

            let handle = s.spawn(move || {
                execute_single_copy(&task, &src, &dst, &db, &logger, &sync_stamp);
            });
            handles.push(handle);
        }
        for h in handles {
            h.join().ok();
        }
    });
}

fn find_peer<'a>(peers: &'a [Arc<ConnectedPeer>], name: &str) -> Option<&'a Arc<ConnectedPeer>> {
    peers.iter().find(|p| p.name == name)
}

fn execute_single_copy(
    task: &CopyTask,
    src: &ConnectedPeer,
    dst: &ConnectedPeer,
    db: &Database,
    logger: &Logger,
    sync_stamp: &str,
) {
    let xfer_ts = timestamp::now();
    let xfer_uuid = uuid::Uuid::new_v4().to_string();
    let basename = hash::basename(&task.path);
    let parent = path_parent(&task.path);

    let xfer_dir = format!(
        "{}/.kitchensync/XFER/{}/{}",
        if parent.is_empty() { "." } else { &parent },
        xfer_ts,
        xfer_uuid
    );
    let xfer_path = format!("{}/{}", xfer_dir, basename);

    // Acquire connections from both pools
    let src_guard = match src.pool.acquire() {
        Ok(g) => g,
        Err(e) => {
            logger.error(&format!("Copy failed {} (src pool): {}", task.path, e));
            return;
        }
    };
    let dst_guard = match dst.pool.acquire() {
        Ok(g) => g,
        Err(e) => {
            logger.error(&format!("Copy failed {} (dst pool): {}", task.path, e));
            return;
        }
    };

    // Step 1: Pipelined transfer to XFER staging
    let (tx, rx): (Sender<Vec<u8>>, Receiver<Vec<u8>>) = bounded(CHANNEL_BOUND);

    let transfer_result = thread::scope(|s| {
        let path = task.path.clone();
        let xfer_path_clone = xfer_path.clone();

        // Reader task: read from source, push chunks to channel
        let reader_handle = s.spawn(move || -> Result<(), PeerError> {
            let mut channel_writer = ChannelWriter { tx };
            src_guard.conn().read_file_to(&path, &mut channel_writer)?;
            drop(channel_writer);
            Ok(())
        });

        // Writer task: pull chunks from channel, write to destination XFER
        let writer_handle = s.spawn(move || -> Result<(), PeerError> {
            let mut channel_reader = ChannelReader {
                rx,
                buf: Vec::new(),
                pos: 0,
            };
            dst_guard.conn().write_file_from(&xfer_path_clone, &mut channel_reader)?;
            Ok(())
        });

        let read_result = reader_handle
            .join()
            .unwrap_or(Err(PeerError::IoError("reader panicked".into())));
        let write_result = writer_handle
            .join()
            .unwrap_or(Err(PeerError::IoError("writer panicked".into())));

        if let Err(e) = read_result {
            return Err(e);
        }
        if let Err(e) = write_result {
            return Err(e);
        }
        Ok(())
    });

    if let Err(e) = transfer_result {
        logger.error(&format!("Transfer failed {}: {}", task.path, e));
        // Clean up XFER staging
        if let Ok(g) = dst.pool.acquire() {
            let _ = g.conn().delete_file(&xfer_path);
            let _ = cleanup_empty_dirs(g.conn(), &xfer_dir);
        }
        return;
    }

    // Steps 2-5: displace existing, swap, set mod_time, cleanup
    if let Ok(g) = dst.pool.acquire() {
        // Step 2: If destination already has a file, displace to BACK
        if let Ok(Some(_)) = g.conn().stat(&task.path) {
            let back_ts = timestamp::now();
            let back_dir = format!(
                "{}/.kitchensync/BACK/{}",
                if parent.is_empty() { "." } else { &parent },
                back_ts
            );
            let back_path = format!("{}/{}", back_dir, basename);
            if let Err(e) = g.conn().rename(&task.path, &back_path) {
                logger.error(&format!("Displacement failed {} -> BACK: {}", task.path, e));
                let _ = g.conn().delete_file(&xfer_path);
                let _ = cleanup_empty_dirs(g.conn(), &xfer_dir);
                return;
            }
        }

        // Step 3: Swap — rename from XFER to final path
        if let Err(e) = g.conn().rename(&xfer_path, &task.path) {
            logger.error(&format!("Swap failed {} from XFER: {}", task.path, e));
            return;
        }

        // Step 4: Set mod_time
        if let Err(e) = g.conn().set_mod_time(&task.path, &task.src_mod_time) {
            logger.debug(&format!("set_mod_time warning {}: {}", task.path, e));
        }

        // Step 5: Clean up empty XFER directories
        let _ = cleanup_empty_dirs(g.conn(), &xfer_dir);
    }

    // Update snapshot: set last_seen after completed copy
    let id = hash::path_hash(&task.path);
    db.set_last_seen(&id, &task.dst_peer, sync_stamp);

    logger.trace(&format!(
        "Copy completed: {} -> {} for {}",
        task.src_peer, task.dst_peer, task.path
    ));
}

/// Displace an entry to BACK/ directory. Used inline during traversal.
pub fn displace_to_back(
    conn: &dyn PeerFs,
    path: &str,
    _logger: &Logger,
) -> Result<(), PeerError> {
    let basename = hash::basename(path);
    let parent = path_parent(path);
    let back_ts = timestamp::now();
    let back_dir = format!(
        "{}/.kitchensync/BACK/{}",
        if parent.is_empty() { "." } else { &parent },
        back_ts
    );
    let back_path = format!("{}/{}", back_dir, basename);
    conn.rename(path, &back_path)?;
    Ok(())
}

fn path_parent(path: &str) -> String {
    let trimmed = path.trim_end_matches('/');
    match trimmed.rfind('/') {
        Some(i) => trimmed[..i].to_string(),
        None => String::new(),
    }
}

fn cleanup_empty_dirs(conn: &dyn PeerFs, dir: &str) -> Result<(), PeerError> {
    let _ = conn.delete_dir(dir);
    let parent = path_parent(dir);
    if !parent.is_empty() {
        let _ = conn.delete_dir(&parent);
    }
    Ok(())
}

struct ChannelWriter {
    tx: Sender<Vec<u8>>,
}

impl Write for ChannelWriter {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        self.tx
            .send(buf.to_vec())
            .map_err(|_| std::io::Error::new(std::io::ErrorKind::BrokenPipe, "channel closed"))?;
        Ok(buf.len())
    }

    fn flush(&mut self) -> std::io::Result<()> {
        Ok(())
    }
}

struct ChannelReader {
    rx: Receiver<Vec<u8>>,
    buf: Vec<u8>,
    pos: usize,
}

impl Read for ChannelReader {
    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
        if self.pos < self.buf.len() {
            let n = std::cmp::min(buf.len(), self.buf.len() - self.pos);
            buf[..n].copy_from_slice(&self.buf[self.pos..self.pos + n]);
            self.pos += n;
            return Ok(n);
        }

        match self.rx.recv() {
            Ok(chunk) => {
                let n = std::cmp::min(buf.len(), chunk.len());
                buf[..n].copy_from_slice(&chunk[..n]);
                if n < chunk.len() {
                    self.buf = chunk;
                    self.pos = n;
                } else {
                    self.buf.clear();
                    self.pos = 0;
                }
                Ok(n)
            }
            Err(_) => Ok(0),
        }
    }
}

File: code/internal/sync/copy.go
package sync

import (
	"fmt"
	"io"
	"path"

	"github.com/google/uuid"

	"kitchensync/internal/log"
	"kitchensync/internal/peer"
	"kitchensync/internal/pool"
	"kitchensync/internal/snapshot"
	"kitchensync/internal/timestamp"
)

// CopyFile copies a file from src to dst peer using their connection pools.
func CopyFile(srcPeer, dstPeer *peer.Peer, relPath string, modTime string, byteSize int64) error {
	srcConn, dstConn, err := pool.AcquireOrdered(srcPeer.Pool, dstPeer.Pool)
	if err != nil {
		return fmt.Errorf("acquire connections: %w", err)
	}
	defer srcPeer.Pool.Release(srcConn)
	defer dstPeer.Pool.Release(dstConn)

	ts := timestamp.FormatTime(timestamp.Now())
	basename := path.Base(relPath)
	parentDir := path.Dir(relPath)
	if parentDir == "." {
		parentDir = ""
	}
	tmpBase := path.Join(parentDir, ".kitchensync/TMP")
	tmpDir := path.Join(tmpBase, ts, uuid.New().String())
	tmpPath := path.Join(tmpDir, basename)

	log.Trace("pipe reader-start src=%s file=%s", srcPeer.Label(), relPath)
	log.Trace("pipe writer-start dst=%s file=%s", dstPeer.Label(), relPath)

	// Pipelined transfer using a pipe
	pr, pw := io.Pipe()
	errCh := make(chan error, 2)

	// Reader goroutine
	go func() {
		reader, err := srcConn.ReadFile(relPath)
		if err != nil {
			pw.CloseWithError(err)
			errCh <- err
			return
		}
		_, err = io.Copy(pw, reader)
		reader.Close()
		pw.CloseWithError(err)
		errCh <- err
		log.Trace("pipe reader-done  src=%s file=%s", srcPeer.Label(), relPath)
	}()

	// Writer goroutine
	go func() {
		err := dstConn.WriteFile(tmpPath, pr)
		if err != nil {
			// Close the pipe reader so the reader goroutine unblocks if WriteFile
			// returned before consuming all data (e.g., MkdirAll or Create failed).
			pr.CloseWithError(err)
		}
		errCh <- err
		log.Trace("pipe writer-done  dst=%s file=%s", dstPeer.Label(), relPath)
	}()

	// Wait for both
	err1 := <-errCh
	err2 := <-errCh
	if err1 != nil || err2 != nil {
		// Clean up TMP
		dstConn.DeleteFile(tmpPath)
		cleanupEmptyParents(dstConn, tmpDir, tmpBase)
		if err1 != nil {
			return err1
		}
		return err2
	}

	// Displace existing file at target to BAK/
	if exists, _ := dstConn.Exists(relPath); exists {
		bakTs := timestamp.FormatTime(timestamp.Now())
		bakPath := path.Join(parentDir, ".kitchensync/BAK", bakTs, basename)
		if err := dstConn.Rename(relPath, bakPath); err != nil {
			log.Error("displace existing at %s on %s: %v", relPath, dstPeer.Label(), err)
			// Clean up TMP and abort
			dstConn.DeleteFile(tmpPath)
			cleanupEmptyParents(dstConn, tmpDir, tmpBase)
			return err
		}
	}

	// Atomic swap: rename TMP -> final
	if err := dstConn.Rename(tmpPath, relPath); err != nil {
		log.Error("rename tmp to final at %s on %s: %v", relPath, dstPeer.Label(), err)
		dstConn.DeleteFile(tmpPath)
		cleanupEmptyParents(dstConn, tmpDir, tmpBase)
		return err
	}

	// Set mod_time to winning mod_time
	mt, _ := snapshot.ParseModTime(modTime)
	if err := dstConn.SetModTime(relPath, mt); err != nil {
		log.Warn("set mod_time on %s at %s: %v", relPath, dstPeer.Label(), err)
	}

	// Best-effort permission copy
	if perm, err := srcConn.GetPermissions(relPath); err == nil {
		dstConn.SetPermissions(relPath, perm)
	}

	// Clean up empty TMP dirs
	cleanupEmptyParents(dstConn, tmpDir, tmpBase)

	// Post-copy snapshot update: set last_seen = now
	nowStr := snapshot.NowStr()
	dstPeer.Snap.SetLastSeen(relPath, nowStr)

	return nil
}

func cleanupEmptyParents(fs interface{ DeleteDir(string) error }, dir, stopAt string) {
	for dir != stopAt && dir != "" && dir != "." {
		fs.DeleteDir(dir)
		dir = path.Dir(dir)
	}
}
